eth/gasprice/feehistory.go

```diff
@@ -243,24 +238,37 @@ func (oracle *Oracle) FeeHistory(ctx context.Context, blocks int, unresolvedLast
 					return
 				}
 
+				var pending bool
+				cacheKey := struct {
+					number      uint64
+					percentiles string
+				}{blockNumber, string(percentileKey)}
 				fees := &blockFees{blockNumber: blockNumber}
 				if pendingBlock != nil && blockNumber >= pendingBlock.NumberU64() {
 					fees.block, fees.receipts = pendingBlock, pendingReceipts
+					pending = true
 				} else {
-					if len(rewardPercentiles) != 0 {
-						fees.block, fees.err = oracle.backend.BlockByNumber(ctx, rpc.BlockNumber(blockNumber))
-						if fees.block != nil && fees.err == nil {
-							fees.receipts, fees.err = oracle.backend.GetReceipts(ctx, fees.block.Hash())
-						}
+					if p, ok := oracle.historyCache.Get(cacheKey); ok {
+						fees.results = p.(processedFees)
 					} else {
-						fees.header, fees.err = oracle.backend.HeaderByNumber(ctx, rpc.BlockNumber(blockNumber))
+						if len(rewardPercentiles) != 0 {
+							fees.block, fees.err = oracle.backend.BlockByNumber(ctx, rpc.BlockNumber(blockNumber))
+							if fees.block != nil && fees.err == nil {
+								fees.receipts, fees.err = oracle.backend.GetReceipts(ctx, fees.block.Hash())
+							}
+						} else {
+							fees.header, fees.err = oracle.backend.HeaderByNumber(ctx, rpc.BlockNumber(blockNumber))
+						}
 					}
 				}
 				if fees.block != nil {
 					fees.header = fees.block.Header()
 				}
 				if fees.header != nil {
 					oracle.processBlock(fees, rewardPercentiles)
+					if fees.err == nil && !pending {
```

Comment:
nitpicks, it's not obvious that logic is using the cached result or process to get the result.
Can we send back the result by `results <- fees` immediately if we find the result in Cache? 
So that the logic is more clear.